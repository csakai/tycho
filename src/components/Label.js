import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired
  }

  getPosition = (pos) => {
    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }
  
  render() {
    return (
      <span style={this.getPosition(this.props.position)}>
        {this.props.text}
      </span>
    );
  }
}

export default Label;
