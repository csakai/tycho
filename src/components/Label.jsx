import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    active: PropTypes.bool
  }

  getPosition = (pos) => {
    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }

  getChildren = (children, active) => {
    if (active) {
      return children;
    }
    return null;
  }
  
  render() {
    return (
      <div>
        <span style={this.getPosition(this.props.position)}>
          {this.props.text}
        </span>
        {this.getChildren(this.props.children, this.props.active)}
      </div>
    );
  }
}

export default Label;
