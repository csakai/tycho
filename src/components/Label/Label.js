import React from 'react';
import PropTypes from 'prop-types';

class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    orbital: PropTypes.object.isRequired
  }

  getPosition = () => {
    const pos = this.props.position;

    if (pos) {
      return {
        position: 'absolute',
        top: `${pos.top}px`,
        left: `${pos.left}px`
      };
    }
    return null;
  }
  
  render() {
    return (
      <span style={this.getPosition()}>
        {this.props.orbital.name}
      </span>
    );
  }
}

export default Label;
