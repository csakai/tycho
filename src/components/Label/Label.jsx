import React from 'react';
import PropTypes from 'prop-types';

export default class Label extends React.Component {

  static propTypes = {
    position: PropTypes.object,
    text: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
    active: PropTypes.bool,
    enabled: PropTypes.bool
  }
  
  render() {
    return (
      <div className="label" style={{
        display: this.props.enabled ? 'block' : 'none'
      }}>
        <span
          className="label__text"
          style={this.props.position}
          onClick={this.props.onClick}>
          {!this.props.active && this.props.text}
        </span>
        {this.props.active && this.props.children}
      </div>
    );
  }
}
