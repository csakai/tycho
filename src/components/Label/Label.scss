.label {
  &__text {
    position: absolute;
    text-shadow: 1px 1px 1px #c0c0c0;
    opacity: 0.75;
    font-family: Georgia, serif;
    cursor: pointer;

    &:hover {
      opacity: 1;
    }
  }
}
