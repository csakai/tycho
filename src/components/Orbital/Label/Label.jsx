import React from 'react';
import PropTypes from 'prop-types';
import DomLabel from 'three-dom-label';

export default class Label extends React.Component {

    static propTypes = {
        text: PropTypes.string.isRequired,
        id: PropTypes.string.isRequired,
        action: PropTypes.object.isRequired,
        targetId: PropTypes.string
    }

    componentDidMount = () => {
        this.label = this.getLabel();   
        this.refs.group.add(this.label);
    }

    componentWillUnmount = () => this.label.unmount()

    componentWillReceiveProps = (nextProps) => {
        this.maybeUpdateClassName(nextProps);
    }

    /**
     * Updates class name if the active target has changed.
     * 
     * @param {Object} props
     * @param {String} props.id - id of current orbital
     * @param {String} props.targetId - id of active orbital
     */
    maybeUpdateClassName = ({id, targetId}) => {
        if (this.props.targetId !== targetId) {
            const isActive = id === targetId;
            const classSuffix = isActive ? 'active' : 'inactive';

            this.label.setClass(`label label--${classSuffix}`);
        }
    }

    /**
     * Renders a new label, with mouse events bound to it.
     * 
     * @returns {Object3D} label
     */
    getLabel = () => {
        const {id, text, action} = this.props;
        const classPrefix = `label label`;

        return new DomLabel({
            ...this.props,
            events: {
                click: action.setActiveOrbital.bind(this, id, text),
                mouseover: action.addHighlightedOrbital.bind(this, id),
                mouseout: action.removeHighlightedOrbital.bind(this, id)
            },
            className: `${classPrefix}--inactive`
        });
    }

    render() {
        return <group ref="group"></group>;
    }
}