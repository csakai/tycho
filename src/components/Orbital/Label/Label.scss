@import 'globals.scss';

.label {
    &--active {
        animation: glowing-text $transition-speed infinite;
        pointer-events: none;
        font-size: $font-size-medium;
    }

    &--inactive {
        color: $brand-primary;
        cursor: pointer;
        transition: font-size $transition-speed;

        &:hover {
            text-shadow: $text-shadow;
            font-size: $font-size-medium;
        }
    }
}

@keyframes glowing-text {
  0%   { text-shadow: 0 }
  40%  { text-shadow: $text-shadow }
  60%  { text-shadow: $text-shadow }
  100% { text-shadow: 0 }
}