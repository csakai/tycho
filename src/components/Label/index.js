import Label from './Label';
import './Label.css';

export default Label;
