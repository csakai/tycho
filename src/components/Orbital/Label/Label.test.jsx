import React from 'react';
import toJson from 'enzyme-to-json';
import { shallow } from 'enzyme';
import Label from './Label';

jest.mock('three-dom-label', () => {
    return function (settings) {
        this.unmount = jest.fn();
        this.setClass = jest.fn();

        Object.assign(this, settings);
    }
});

describe.only('Orbital Label Component', () => {
    let component, label;

    beforeEach(() => {
        component = shallow(
            <Label
                action={{
                    setActiveOrbital: jest.fn(),
                    addHighlightedOrbital: jest.fn(),
                    removeHighlightedOrbital: jest.fn()
                }}
                text="Test Planet"
                id="testPlanet"
            />
        );
        label = component.instance();
    });

    describe('componentDidMount()', () => {
        beforeEach(() => {
            label.refs = {
                group: {
                    add: jest.fn()
                }
            };
        });

        it('should assign a new label', () => {
            label.componentDidMount();
            expect(label).toHaveProperty('label');
        });

        it('should add the label to the group', () => {
            const spy = jest.spyOn(label.refs.group, 'add');

            label.componentDidMount();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledWith(label.label);
            expect(spy).toHaveBeenCalledTimes(1);
        });
    });

    describe('componentWillUnmount()', () => {
        beforeEach(() => {
            label.label = { unmount: jest.fn() };
        });

        it('should unmount the DOM label', () => {
            const spy = jest.spyOn(label.label, 'unmount');

            label.componentWillUnmount();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });
    });

    describe('componentWillReceiveProps()', () => {
        it('should call maybeUpdateClassName() with nextProps', () => {
            const targetId = 'testPlanet';
            const nextProps = { targetId };
            const spy = jest.spyOn(label, 'maybeUpdateClassName');

            label.props = { targetId };
            label.componentWillReceiveProps(nextProps);

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(nextProps);
        });
    });

    describe('maybeUpdateClassName()', () => {
        let spy;
        const id = 'Earth';
        const targetId = 'Mars';

        beforeEach(() => {
            label.props = { targetId };
            label.label = {
                setClass: jest.fn()
            };
            spy = jest.spyOn(label.label, 'setClass');
        });

        describe('when the targetId has changed', () => {
            it('should set the class suffix to \'inactive\' when the targetId does not match the label id', () => {
                label.maybeUpdateClassName({id, targetId: 'Jupiter'});

                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledTimes(1);
                expect(spy).toHaveBeenCalledWith('label label--inactive');
            });

            it('should set the class suffix to \'active\' when the targetId matches the label id', () => {
                label.maybeUpdateClassName({id, targetId: id});

                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledTimes(1);
                expect(spy).toHaveBeenCalledWith('label label--active');
            });
        });

        describe('when the targetId has not changed', () => {
            it('should not change the class name', () => {
                label.maybeUpdateClassName({ id, targetId });

                expect(spy).not.toHaveBeenCalled();
            });
        });
    });

    describe('getLabel()', () => {
        let result;
        const text = "Test Label";
        const id = "testPlanet";

        beforeEach(() => {
            const action = {
                setActiveOrbital: jest.fn(),
                addHighlightedOrbital: jest.fn(),
                removeHighlightedOrbital: jest.fn()
            };

            label.props = { text, id, action };
            result = label.getLabel();
        });

        it('should pass the text property to the label', () => {

            expect(result).toHaveProperty('text');
            expect(result.text).toEqual(text);
        });

        it('should pass the id property to the label', () => {
            expect(result).toHaveProperty('id');
            expect(result.id).toEqual(id);
        });

        it('should set the className to the \'inactive\' label class', () => {
            expect(result).toHaveProperty('className');
            expect(result.className).toEqual('label label--inactive');
        });
    });

    it('should render the markdown container successfully', () => {
        expect(toJson(component)).toMatchSnapshot();
    });
});