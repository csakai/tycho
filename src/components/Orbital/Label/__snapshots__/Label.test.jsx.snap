// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orbital Label Component should render the markdown container successfully 1`] = `<group />`;
